src/middlewared/middlewared/plugins/vm.py

```diff
@@ -7,481 +6,609 @@
 
+        vm_xml = etree.tostring(self.construct_xml()).decode()
+        if not self.connection.defineXML(vm_xml):
+            raise CallError(f'Unable to define persistent domain for {self.libvirt_domain_name}')
 
-class VMSupervisor(object):
+        self.domain = self.connection.lookupByName(self.libvirt_domain_name)
 
-    def __init__(self, manager, vm):
-        self.manager = manager
+    def undefine_domain(self):
+        if self.domain.isActive():
+            raise CallError(f'Domain {self.libvirt_domain_name} is active. Please stop it first')
 
-        os.makedirs('/var/log/vm', exist_ok=True)
-        self.logger = self.manager.logger.getChild(f'vm_{vm["id"]}')
+        if self.vm_data['bootloader'] == 'GRUB':
+            shutil.rmtree(
+                os.path.join('/tmp/grub', self.libvirt_domain_name), ignore_errors=True
+            )
+
+        self.domain.undefine()
+        self.domain = None
+
+    def __getattribute__(self, item):
+        retrieved_item = object.__getattribute__(self, item)
+        if callable(retrieved_item) and item in ('start', 'stop', 'restart', 'poweroff', 'undefine_domain', 'status'):
+            if not getattr(self, 'domain', None):
+                raise CallError('Domain attribute not defined, please re-instantiate the VM class')
+
+        return retrieved_item
+
+    def start(self, vm_data=None):
+        if self.domain.isActive():
+            raise CallError(f'{self.libvirt_domain_name} domain is already active')
+
+        self.update_domain(vm_data)
 
-        handler = middlewared.logger.ErrorProneRotatingFileHandler(
-            f'/var/log/vm/{vm["name"]}_{vm["id"]}',
-            maxBytes=10485760,
-            backupCount=5,
+        # Let's ensure that we are able to boot a GRUB based VM
+        if self.vm_data['bootloader'] == 'GRUB' and not any(
+            isinstance(d, RAW) and d.data['attributes'].get('boot') for d in self.devices
+        ):
+            raise CallError(f'Unable to find boot devices for {self.libvirt_domain_name} domain')
+
+        for device in self.devices:
+            device.pre_start_vm()
+
+        if self.domain.create() < 0:
+            raise CallError(f'Failed to boot {self.vm_data["name"]} domain')
+
+        for device in self.devices:
+            device.post_start_vm()
+
+    def _before_stopping_checks(self):
+        if not self.domain.isActive():
+            raise CallError(f'{self.libvirt_domain_name} domain is not active')
+
+    def stop(self):
+        self._before_stopping_checks()
+
+        for device in self.devices:
+            device.pre_stop_vm()
+
+        self.domain.shutdown()
+
+        # We wait for 30 seconds before initiating post stop activities for the vm
+        # This is done because the shutdown call above is non-blocking
+        shutdown_timeout = 30
+        while shutdown_timeout > 0 and self.status()['state'] == 'RUNNING':
+            shutdown_timeout -= 5
+            time.sleep(5)
+
```

Comment:
Why is there no check that machine stop is successful (and in restart method there is)?

Reply:
My thinking here was that a shutdown signal is being sent to the VM, it's possible that there is some blocking task which is forcing it to not go in shutdown mode. In that case user will see that the status of VM is still running and he/she can issue a poweroff. If we make this a blocking call ( the shutdown one to ensure that shutdown happens ), it would have to be a job or if we want to raise an exception, what's the ideal time to wait for the machine to cleanly shutdown ?

Reply:
My thinking is shutdown timeout should be user-configurable.

Also now you are performing `device.post_stop_vm()` even if vm has not stopped which is not good in any case.

Reply:
User configurable as in an attribute specific to a VM or all VM's ?

Reply:
In either cases we would have to come up with a reasonable default as well, do you think 30 seconds is good in that case ?

Reply:
> User configurable as in an attribute specific to a VM or all VM's ?

To `stop` operation.

30 seems to be too short, we had something like 90 in FreeBSD rc scripts?

Reply:
I'll have this updated, also i think we shouldn't run post stop activities if the VM is running in that case. Do you agree ? This can cause issues wrt VM stopping down later and we unable to run post stop for the devices in both poweroff/stop as VM is not active

Reply:
I think i just reversed it, i should run post stop regardless VM is active or not as the user had a chance to configure the timeout for stopping. Thoughts ?

Reply:
Can we spawn a background coroutine that will perform post stop actions when VM is finally shut down? Regardless of stop job status.